File: leetcode/Medium/MergeIntervals.java
package solutions.medium;

import solutions.util.ArrayUtils;

import java.util.*;

// 56 https://leetcode.com/problems/merge-intervals/
public class MergeIntervals {
    public static void main(String[] args) {
        MergeIntervals obj = new MergeIntervals();

        ArrayUtils.print2DArray(obj.merge(new int[][]{
                {1, 4},
                {4, 5}
        }));
    }

    public int[][] mergeUsingStack(int[][] intervals) {
        Arrays.sort(intervals, new Comparator<int[]>() {
            @Override
            public int compare(int[] i1, int[] i2) {
                return i1[0] - i2[0];
            }
        });

        Stack<int[]> stack = new Stack<>();
        stack.push(intervals[0]);

        for (int i = 1; i < intervals.length; i++) {
            int[] top = stack.peek();
            if (top[1] < intervals[i][0]) {
                // non overlapping
                stack.push(intervals[i]);
            } else if (top[1] < intervals[i][1]) {
                top[1] = intervals[i][1];
                stack.pop();
                stack.push(top);
            }
        }
        int[][] result = new int[stack.size()][2];
        int i = 0;
        while (!stack.isEmpty()) {
            result[i++] = stack.pop();
        }

        return result;
    }

    public int[][] merge(int[][] intervals) {
        if (intervals.length < 2) {
            return intervals;
        }

        Arrays.sort(intervals, (i1, i2) -> i1[0] - i2[0]);

        int[] current = intervals[0];
        List<int[]> outputList = new ArrayList<>();
        outputList.add(current);

        for (int[] interval : intervals) {
            int currentEnd = current[1];
            int nextStart = interval[0];
            int nextEnd = interval[1];

            if (currentEnd >= nextStart) {
                current[1] = Math.max(currentEnd, nextEnd);
            } else {
                current = interval;
                outputList.add(current);
            }

        }

        return outputList.toArray(new int[outputList.size()][]);

    }

}

File: techniques/intervals/MergeIntervals.kt
import java.util.*

//56 https://leetcode.com/problems/merge-intervals/
fun merge(intervals: Array<IntArray>): Array<IntArray> {
    Arrays.sort(intervals) { i1, i2 ->
        if (i1[0] == i2[0]) i1[1] - i2[1]
        else i1[0] - i2[0]
    }
    val result = mutableListOf<IntArray>()

    var currentInterval = intervals[0]
    result.add(currentInterval)

    for (interval in intervals) {
        if (currentInterval[1] >= interval[0]) {
            // keep updating the upper bound of interval untill there is an overlap
            // the lower bound need not change because the array is sorted, so it will be minimum from start
            currentInterval[1] = Math.max(currentInterval[1], interval[1])
        } else {
            // We have found a non overlapping interval, so update the current interval and repeat the same process all
            // over again
            currentInterval = interval
            result.add(currentInterval)
        }
    }

    return result.toTypedArray()
}

fun main() {
    val result1 = merge(arrayOf(intArrayOf(1, 3), intArrayOf(2, 6), intArrayOf(8, 10), intArrayOf(15, 18)))
    result1.forEach { print(it.contentToString()) }

    println()

    val result2 = merge(arrayOf(intArrayOf(1, 5), intArrayOf(2, 10), intArrayOf(8, 9)))
    result2.forEach { println(it.contentToString()) }


}